runner/src/main.rs

```diff
@@ -595,19 +653,267 @@ struct Example {
     server: ExampleServer,
     #[serde(default)]
     backends: HashMap<String, Executable>,
-    #[serde(default)]
-    applications: HashMap<String, Application>,
+    applications: Applications,
     clients: HashMap<String, Executable>,
 }
 
+impl Example {
+    fn construct_application_build_steps(&self, application_variant: &str) -> Vec<Step> {
+        self.applications
+            .construct_application_build_steps(application_variant, &self.name)
+    }
+
+    fn construct_example_server_run_step(
+        &self,
+        opt: either::Either<&RunExamples, &RunFunctionsExamples>,
+        run_clients: Step,
+    ) -> Step {
+        self.applications
+            .construct_example_server_run_step(opt, &self, run_clients)
+    }
+
+    fn construct_backend_build_steps(&self, build_client: &BuildClient) -> Vec<Step> {
+        self.backends
+            .iter()
+            .map(move |(name, backend)| Step::Single {
+                name: name.to_string(),
+                command: build(&backend.target, build_client),
+            })
+            .collect()
+    }
+
+    /// Recursively constructs backend steps.
+    fn construct_backend_run_steps(
+        &self,
+        run_server_clients: Step,
+        build_client: &BuildClient,
+    ) -> Step {
+        self.backends
+            .iter()
+            // First iteration includes `run_server_clients` as a foreground step.
+            .fold(run_server_clients, |backend_steps, (name, backend)| {
+                Step::WithBackground {
+                    name: name.to_string(),
+                    // Each `backend` is included as background step.
+                    background: run(&backend, &build_client, Vec::new()),
+                    foreground: Box::new(backend_steps),
+                }
+            })
+    }
+}
+
+/// A construct representing either an Oak Classic or an Oak Functions application.  
+///
+/// The condition that only one of `classic` or `functions` should be non-empty is
+/// checked in each operation of this struct. If neither or both are empty, the
+/// operation panics with an error message.
```

Comment:
I guess the YAML serialization makes it difficult to represent this condition idiomatically as an enum? If that works, it may remove the manual checks and panics below (I think).

Reply:
Yes. There is no default (de)serializer for that kind of enum. But I think we can make it work if we define our own Serializer and Deserializer. I think this can be done in a separate PR, together with some other refactoring.